### PR TITLE
Updated XML Import To Utilize Base64 Decoding On The Body

### PR DESCRIPTION
**Pull request by mhagan-r7**
## In particular
- Added a nil check for xml node importing
- Fixed `web_page` `<body>` tag deserializing to utilize Base64 encoding 

## Verification
- [x] Start `msfconsole`
- [x] Rename `Good_data.txt` to `Good_data.xml`
- [x] Run `db_import Good_data.xml`
- [x] **Verify** the import completes without errors
- [x] Load the `wmap` plugin and display the `web vulns` table
- [x] **Verify** that the `web vulns` table has contents

[Good_data.txt](https://github.com/rapid7/metasploit-framework/files/5617268/Good_data.txt)

### CODE REVIEW

**Comment by gwillcox-r7** [comment]
Looks like this is still crashing if the host is using a PostgreSQL connection:

```
 ~/git/metasploit-framework │ land-pr14445:pr/14445 ?2  ./msfconsole                                                ✔ │ 3m 10s │ 2.7.2 Ruby 
                                                  
  +-------------------------------------------------------+
  |  METASPLOIT by Rapid7                                 |
  +---------------------------+---------------------------+
  |      __________________   |                           |
  |  ==c(______(o(______(_()  | |""""""""""""|======[***  |
  |             )=\           | |  EXPLOIT   \            |
  |            // \\          | |_____________\_______    |
  |           //   \\         | |==[msf >]============\   |
  |          //     \\        | |______________________\  |
  |         // RECON \\       | \(@)(@)(@)(@)(@)(@)(@)/   |
  |        //         \\      |  *********************    |
  +---------------------------+---------------------------+
  |      o O o                |        \'\/\/\/'/         |
  |              o O          |         )======(          |
  |                 o         |       .'  LOOT  '.        |
  | |^^^^^^^^^^^^^^|l___      |      /    _||__   \       |
  | |    PAYLOAD     |""\___, |     /    (_||_     \      |
  | |________________|__|)__| |    |     __||_)     |     |
  | |(@)(@)"""**|(@)(@)**|(@) |    "       ||       "     |
  |  = = = = = = = = = = = =  |     '--------------'      |
  +---------------------------+---------------------------+


       =[ metasploit v6.0.2-dev-a06f605675                ]
+ -- --=[ 2053 exploits - 1109 auxiliary - 347 post       ]
+ -- --=[ 562 payloads - 45 encoders - 10 nops            ]
+ -- --=[ 7 evasion                                       ]

Metasploit tip: Enable verbose logging with set VERBOSE true

msf6 > vulns

Vulnerabilities
===============

Timestamp  Host  Name  References
---------  ----  ----  ----------

msf6 > db_status
[*] Connected to remote_data_service: (https://localhost:5443). Connection type: http. Connection name: local-https-data-service.
msf6 > db_
db_connect        db_export         db_nmap           db_remove         db_status         
db_disconnect     db_import         db_rebuild_cache  db_save           
msf6 > db_disconnect
Successfully disconnected from the data service: remote_data_service: (https://localhost:5443).
msf6 > db_status
[*] Connected to msf. Connection type: postgresql.
msf6 > vulns

Vulnerabilities
===============

Timestamp  Host  Name  References
---------  ----  ----  ----------

msf6 > db_remove
Usage: db_remove <name>

Delete the specified saved data service.

msf6 > db_rebuild_cache 
This command is deprecated with Metasploit 5
msf6 > db_import Good_data.xml
[*] Importing 'Metasploit XML' data
[*] Importing host 192.168.0.3
[-] Error while running command db_import: Validation failed: Params is not an Array.  Valid parameters are an Array<Array(String, String)>.

Call stack:
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/validations.rb:80:in `raise_validation_error'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/validations.rb:52:in `save!'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/transactions.rb:315:in `block in save!'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/transactions.rb:387:in `block in with_transaction_returning_status'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/database_statements.rb:267:in `block in transaction'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/transaction.rb:239:in `block in within_new_transaction'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/transaction.rb:236:in `synchronize'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/transaction.rb:236:in `within_new_transaction'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/database_statements.rb:267:in `transaction'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/transactions.rb:212:in `transaction'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/transactions.rb:385:in `with_transaction_returning_status'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/transactions.rb:315:in `save!'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/suppressor.rb:48:in `save!'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/web.rb:366:in `block in report_web_vuln'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:416:in `with_connection'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/web.rb:291:in `report_web_vuln'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/import/metasploit_framework/xml.rb:651:in `import_msf_web_element'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/import/metasploit_framework/xml.rb:201:in `import_msf_web_vuln_element'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/import/metasploit_framework/xml.rb:261:in `block in import_msf_xml'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/nokogiri-1.10.10/lib/nokogiri/xml/reader.rb:107:in `each'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/import/metasploit_framework/xml.rb:252:in `import_msf_xml'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/import.rb:100:in `import'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/import.rb:219:in `import_file'
/home/gwillcox/git/metasploit-framework/lib/metasploit/framework/data_service/proxy/db_import_data_proxy.rb:17:in `block in import_file'
/home/gwillcox/git/metasploit-framework/lib/metasploit/framework/data_service/proxy/core.rb:166:in `data_service_operation'
/home/gwillcox/git/metasploit-framework/lib/metasploit/framework/data_service/proxy/db_import_data_proxy.rb:15:in `import_file'
/home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1464:in `block (3 levels) in cmd_db_import'
/home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1457:in `each'
/home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1457:in `block (2 levels) in cmd_db_import'
/home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1451:in `each'
/home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1451:in `block in cmd_db_import'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:416:in `with_connection'
/home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1446:in `cmd_db_import'
/home/gwillcox/git/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:523:in `run_command'
/home/gwillcox/git/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:474:in `block in run_single'
/home/gwillcox/git/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:468:in `each'
/home/gwillcox/git/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:468:in `run_single'
/home/gwillcox/git/metasploit-framework/lib/rex/ui/text/shell.rb:158:in `run'
/home/gwillcox/git/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
/home/gwillcox/git/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'
./msfconsole:23:in `<main>'
msf6 > 
```

**Comment by gwillcox-r7** [comment]
Alternatively if the host is using the `http` connection type the data imports fine but I don't see any results in the `vulns` output and I'm not sure if this is intentional or not:

```
 ~/git/metasploit-framework │ land-pr14445:pr/14445 ?2  ./msfconsole                                                ✔ │ 5m 28s │ 2.7.2 Ruby 
                                                  
               .;lxO0KXXXK0Oxl:.
           ,o0WMMMMMMMMMMMMMMMMMMKd,
        'xNMMMMMMMMMMMMMMMMMMMMMMMMMWx,
      :KMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMK:
    .KMMMMMMMMMMMMMMMWNNNWMMMMMMMMMMMMMMMX,
   lWMMMMMMMMMMMXd:..     ..;dKMMMMMMMMMMMMo
  xMMMMMMMMMMWd.               .oNMMMMMMMMMMk
 oMMMMMMMMMMx.                    dMMMMMMMMMMx
.WMMMMMMMMM:                       :MMMMMMMMMM,
xMMMMMMMMMo                         lMMMMMMMMMO
NMMMMMMMMW                    ,cccccoMMMMMMMMMWlccccc;
MMMMMMMMMX                     ;KMMMMMMMMMMMMMMMMMMX:
NMMMMMMMMW.                      ;KMMMMMMMMMMMMMMX:
xMMMMMMMMMd                        ,0MMMMMMMMMMK;
.WMMMMMMMMMc                         'OMMMMMM0,
 lMMMMMMMMMMk.                         .kMMO'
  dMMMMMMMMMMWd'                         ..
   cWMMMMMMMMMMMNxc'.                ##########
    .0MMMMMMMMMMMMMMMMWc            #+#    #+#
      ;0MMMMMMMMMMMMMMMo.          +:+
        .dNMMMMMMMMMMMMo          +#++:++#+
           'oOWMMMMMMMMo                +:+
               .,cdkO0K;        :+:    :+:                                
                                :::::::+:
                      Metasploit

       =[ metasploit v6.0.2-dev-a06f605675                ]
+ -- --=[ 2053 exploits - 1109 auxiliary - 347 post       ]
+ -- --=[ 562 payloads - 45 encoders - 10 nops            ]
+ -- --=[ 7 evasion                                       ]

Metasploit tip: You can use help to view all available commands

msf6 > db_status
[*] Connected to remote_data_service: (https://localhost:5443). Connection type: http. Connection name: local-https-data-service.
msf6 > db_import Good_data.xml
[*] Successfully imported /home/gwillcox/git/metasploit-framework/Good_data.xml
msf6 > vulns

Vulnerabilities
===============

Timestamp  Host  Name  References
---------  ----  ----  ----------

msf6 > hosts

Hosts
=====

address      mac  name         os_name  os_flavor  os_sp  purpose  info  comments
-------      ---  ----         -------  ---------  -----  -------  ----  --------
127.0.0.1                                                                
192.168.0.3       192.168.0.3  Linux                      server         

msf6 > loot

Loot
====

host       service  type      name  content     info  path
----       -------  ----      ----  -------     ----  ----
127.0.0.1           password        text/plain        /home/gwillcox/.msf4/loot/4b2ca63364c94d69e846-20201116174509_default_127.0.0.1_password_392671.txt

msf6 > notes

Notes
=====

 Time  Host  Service  Port  Protocol  Type  Data
 ----  ----  -------  ----  --------  ----  ----

msf6 > services
Services
========

host         port  proto  name  state  info
----         ----  -----  ----  -----  ----
192.168.0.3  80    tcp    http  open   Apache/2.2.8 (Ubuntu) DAV/2

msf6 > analyze
[*] Analyzing 127.0.0.1...
[*] No suggestions for 127.0.0.1.
[*] Analyzing 192.168.0.3...
[*] No suggestions for 192.168.0.3.
msf6 > 

```

**Comment by acammack-r7** [comment]
@gwillcox-r7 web vulns can be viewed with `wmap_vulns -l` from the `wmap` plugin. I get the same stack trace you do still, but afterwards I do get some new web vulns:
```
msf6 > db_import web_import.xml
[*] Importing 'Metasploit XML' data
[*] Importing host 192.168.0.3
[-] Error while running command db_import: Validation failed: Params is not an Array.  Valid parameters are an Array<Array(String, String)>.
... snip ...

msf6 > load wmap 

.-.-.-..-.-.-..---..---.
| | | || | | || | || |-'
`-----'`-'-'-'`-^-'`-'
[WMAP 1.5.1] ===  et [  ] metasploit.com 2012
[*] Successfully loaded plugin: wmap
msf6 > wmap_vulns -l
[*] + [192.168.0.3] (192.168.0.3): Publicly-Writable-Directory /dav/
[*]     Publicly writable directory (HTTP PUT) The logged server directory is publicly writable.
[*]     GET wYHYIXBUu3
[*] + [192.168.0.3] (192.168.0.3): Publicly-Writable-Directory /dav/docs/CHANGELOG
[*]     Publicly writable directory (HTTP PUT) The logged server directory is publicly writable.
[*]     GET LDL0dbsxgd
```

File: lib/msf/core/db_manager/import/metasploit_framework/xml.rb [hunk 1]
```diff
@@ -164,8 +165,18 @@ def import_msf_web_page_element(element, options={}, &notifier)
           element.at('headers'),
           options[:allow_yaml]
       )
-      info[:headers] = nils_for_nulls(unserialized_headers)
 
+      unserialized_body = unserialize_object(element.at('body'), options[:allow_yaml])
+      unless unserialized_body.blank?
+        begin
+          unserialized_body = Base64.urlsafe_decode64(unserialized_body)
```
Comment (acammack-r7) [inline lib/msf/core/db_manager/import/metasploit_framework/xml.rb:172]:
```suggestion
          unserialized_body = Base64.urlsafe_decode64(unserialized_body).b
```

**Comment by gwillcox-r7** [comment]
Quick update on this, without implementing the change that @acammack-r7 suggested:

```
 ~/git/metasploit-framework │ @a06f6056 ?2  ./msfconsole                                                                     ✔ │ 2.7.2 Ruby 
                                                  

      .:okOOOkdc'           'cdkOOOko:.
    .xOOOOOOOOOOOOc       cOOOOOOOOOOOOx.
   :OOOOOOOOOOOOOOOk,   ,kOOOOOOOOOOOOOOO:
  'OOOOOOOOOkkkkOOOOO: :OOOOOOOOOOOOOOOOOO'
  oOOOOOOOO.MMMM.oOOOOoOOOOl.MMMM,OOOOOOOOo
  dOOOOOOOO.MMMMMM.cOOOOOc.MMMMMM,OOOOOOOOx
  lOOOOOOOO.MMMMMMMMM;d;MMMMMMMMM,OOOOOOOOl
  .OOOOOOOO.MMM.;MMMMMMMMMMM;MMMM,OOOOOOOO.
   cOOOOOOO.MMM.OOc.MMMMM'oOO.MMM,OOOOOOOc
    oOOOOOO.MMM.OOOO.MMM:OOOO.MMM,OOOOOOo
     lOOOOO.MMM.OOOO.MMM:OOOO.MMM,OOOOOl
      ;OOOO'MMM.OOOO.MMM:OOOO.MMM;OOOO;
       .dOOo'WM.OOOOocccxOOOO.MX'xOOd.
         ,kOl'M.OOOOOOOOOOOOO.M'dOk,
           :kk;.OOOOOOOOOOOOO.;Ok:
             ;kOOOOOOOOOOOOOOOk:
               ,xOOOOOOOOOOOx,
                 .lOOOOOOOl.
                    ,dOd,
                      .

       =[ metasploit v6.0.2-dev-a06f605675                ]
+ -- --=[ 2053 exploits - 1109 auxiliary - 347 post       ]
+ -- --=[ 562 payloads - 45 encoders - 10 nops            ]
+ -- --=[ 7 evasion                                       ]

Metasploit tip: To save all commands executed since start up to a file, use the makerc command

msf6 > db_status
[*] Connected to remote_data_service: (https://localhost:5443). Connection type: http. Connection name: local-https-data-service.
msf6 > db_import Good_data.xml
[*] Successfully imported /home/gwillcox/git/metasploit-framework/Good_data.xml
msf6 > load wmap

.-.-.-..-.-.-..---..---.
| | | || | | || | || |-'
`-----'`-'-'-'`-^-'`-'
[WMAP 1.5.1] ===  et [  ] metasploit.com 2012
[*] Successfully loaded plugin: wmap
msf6 > wmap_vulns
[*] Usage: wmap_vulns [options]
	-h 		Display this help text
	-l 		Display web vulns table

msf6 > wmap_vulns -l
[*] + [192.168.0.3] (192.168.0.3): Publicly-Writable-Directory /dav/
[*] 	Publicly writable directory (HTTP PUT) The logged server directory is publicly writable.
[*] 	GET wYHYIXBUu3
[*] + [192.168.0.3] (192.168.0.3): Publicly-Writable-Directory /dav/docs/CHANGELOG
[*] 	Publicly writable directory (HTTP PUT) The logged server directory is publicly writable.
[*] 	GET LDL0dbsxgd
msf6 > 
```

So it seems to work with the normal web interface. As for the PostgreSQL database instance, you are right it does import but still experiences that stack trace:

```
 ~/git/metasploit-framework │ @a06f6056 ?2  ./msfconsole                                                               ✔ │ 28s │ 2.7.2 Ruby 
                                                  

 ______________________________________________________________________________
|                                                                              |
|                   METASPLOIT CYBER MISSILE COMMAND V5                        |
|______________________________________________________________________________|
      \                                  /                      /
       \     .                          /                      /            x
        \                              /                      /
         \                            /          +           /
          \            +             /                      /
           *                        /                      /
                                   /      .               /
    X                             /                      /            X
                                 /                     ###
                                /                     # % #
                               /                       ###
                      .       /
     .                       /      .            *           .
                            /
                           *
                  +                       *

                                       ^
####      __     __     __          #######         __     __     __        ####
####    /    \ /    \ /    \      ###########     /    \ /    \ /    \      ####
################################################################################
################################################################################
# WAVE 5 ######## SCORE 31337 ################################## HIGH FFFFFFFF #
################################################################################
                                                           https://metasploit.com


       =[ metasploit v6.0.2-dev-a06f605675                ]
+ -- --=[ 2053 exploits - 1109 auxiliary - 347 post       ]
+ -- --=[ 562 payloads - 45 encoders - 10 nops            ]
+ -- --=[ 7 evasion                                       ]

Metasploit tip: Use the edit command to open the currently active module in your editor

msf6 > load wmap

.-.-.-..-.-.-..---..---.
| | | || | | || | || |-'
`-----'`-'-'-'`-^-'`-'
[WMAP 1.5.1] ===  et [  ] metasploit.com 2012
[*] Successfully loaded plugin: wmap
msf6 > web_vulns -l
[-] Unknown command: web_vulns.
msf6 > wmap_vulns -l
msf6 > db_disconnect 
Successfully disconnected from the data service: remote_data_service: (https://localhost:5443).
msf6 > db_status
[*] Connected to msf. Connection type: postgresql.
msf6 > wmap_vulns -l
msf6 > db_import Good_data.xml
[*] Importing 'Metasploit XML' data
[*] Importing host 192.168.0.3
[-] Error while running command db_import: Validation failed: Params is not an Array.  Valid parameters are an Array<Array(String, String)>.

Call stack:
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/validations.rb:80:in `raise_validation_error'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/validations.rb:52:in `save!'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/transactions.rb:315:in `block in save!'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/transactions.rb:387:in `block in with_transaction_returning_status'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/database_statements.rb:267:in `block in transaction'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/transaction.rb:239:in `block in within_new_transaction'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/transaction.rb:236:in `synchronize'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/transaction.rb:236:in `within_new_transaction'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/database_statements.rb:267:in `transaction'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/transactions.rb:212:in `transaction'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/transactions.rb:385:in `with_transaction_returning_status'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/transactions.rb:315:in `save!'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/suppressor.rb:48:in `save!'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/web.rb:366:in `block in report_web_vuln'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:416:in `with_connection'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/web.rb:291:in `report_web_vuln'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/import/metasploit_framework/xml.rb:651:in `import_msf_web_element'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/import/metasploit_framework/xml.rb:201:in `import_msf_web_vuln_element'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/import/metasploit_framework/xml.rb:261:in `block in import_msf_xml'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/nokogiri-1.10.10/lib/nokogiri/xml/reader.rb:107:in `each'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/import/metasploit_framework/xml.rb:252:in `import_msf_xml'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/import.rb:100:in `import'
/home/gwillcox/git/metasploit-framework/lib/msf/core/db_manager/import.rb:219:in `import_file'
/home/gwillcox/git/metasploit-framework/lib/metasploit/framework/data_service/proxy/db_import_data_proxy.rb:17:in `block in import_file'
/home/gwillcox/git/metasploit-framework/lib/metasploit/framework/data_service/proxy/core.rb:166:in `data_service_operation'
/home/gwillcox/git/metasploit-framework/lib/metasploit/framework/data_service/proxy/db_import_data_proxy.rb:15:in `import_file'
/home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1464:in `block (3 levels) in cmd_db_import'
/home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1457:in `each'
/home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1457:in `block (2 levels) in cmd_db_import'
/home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1451:in `each'
/home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1451:in `block in cmd_db_import'
/home/gwillcox/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:416:in `with_connection'
/home/gwillcox/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1446:in `cmd_db_import'
/home/gwillcox/git/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:523:in `run_command'
/home/gwillcox/git/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:474:in `block in run_single'
/home/gwillcox/git/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:468:in `each'
/home/gwillcox/git/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:468:in `run_single'
/home/gwillcox/git/metasploit-framework/lib/rex/ui/text/shell.rb:158:in `run'
/home/gwillcox/git/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
/home/gwillcox/git/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'
./msfconsole:23:in `<main>'
msf6 > wmap_vulns -l
[*] + [192.168.0.3] (192.168.0.3): Publicly-Writable-Directory /dav/
[*] 	Publicly writable directory (HTTP PUT) The logged server directory is publicly writable.
[*] 	GET wYHYIXBUu3
[*] + [192.168.0.3] (192.168.0.3): Publicly-Writable-Directory /dav/docs/CHANGELOG
[*] 	Publicly writable directory (HTTP PUT) The logged server directory is publicly writable.
[*] 	GET LDL0dbsxgd
msf6 > 
```

**Comment by gwillcox-r7** [comment]
@acammack-r7 Quick question for you but do you have any insight on the stack crash? Was told by Matt that you might be able to provide some insight into why this might be occurring.

**Comment by acammack-r7** [comment]
I don't know why that's popping up. To investigate it, the best thing to do would be to catch that exception in the importer and examine the the invalid data that we are trying to save.

**Comment by gwillcox-r7** [comment]
Okay found the root cause of the issue after digging through the stack traces. Changed the code at `lib/msf/core/db_manager/web.rb` to the following:

```
    vuln = ::Mdm::WebVuln.where(web_site_id: site[:id], path: path, method: meth, pname: pname, name: name, category: cat, query: quer).first_or_initialize
    vuln.name     = name
    vuln.risk     = risk
    vuln.params   = para
    vuln.proof    = proof.to_s
    vuln.category = cat
    vuln.blame    = blame
    vuln.description = desc
    vuln.confidence  = conf
    vuln.payload = payload
    vuln.owner   = owner

    msf_import_timestamps(opts, vuln)
    p vuln
    vuln.save!
```

This showed me the following output right before the crash happened:

```
#<Mdm::WebVuln id: 1, web_site_id: 1, created_at: "2020-11-23 18:15:08", updated_at: "2020-11-23 18:15:08", path: "/dav/", method: "GET", params: [], pname: "path", risk: 5, name: "Publicly writable directory (HTTP PUT)", query: "", category: "Publicly-Writable-Directory", confidence: 100, description: "The logged server directory is publicly writable.", blame: "App Developer", request: nil, proof: "wYHYIXBUu3", owner: nil, payload: "">
#<Mdm::WebVuln id: 2, web_site_id: 1, created_at: "2020-11-23 18:17:46", updated_at: "2020-11-23 18:17:46", path: "/dav/docs/CHANGELOG", method: "GET", params: [], pname: "path", risk: 5, name: "Publicly writable directory (HTTP PUT)", query: "", category: "Publicly-Writable-Directory", confidence: 100, description: "The logged server directory is publicly writable.", blame: "App Developer", request: nil, proof: "LDL0dbsxgd", owner: nil, payload: "">
#<Mdm::WebVuln id: nil, web_site_id: 1, created_at: "2020-11-23 18:24:49", updated_at: "2020-11-23 18:24:49", path: "/mutillidae/", method: "GET", params: "page\n        /etc/passwd", pname: "page", risk: 5, name: "Local File Include", query: "", category: "LFI", confidence: 100, description: "A local file include vulnerability has been identi...", blame: "App Developer", request: nil, proof: "gin Content --> root:x:0:0:root:/root:/bin/bash da...", owner: nil, payload: "">
```

Notice that the final line has the `params` value set to `page\n  /etc/passwd` as a string instead of being the normal array as one would expect. This results in the `activerecord` gem throwing an error as it got data in an incorrect format.

Again none of this is related to the current PR or any of the changes made therein, so I will go ahead and land this for now, however it is something that I think should be addressed in another PR as this seems like some deeper issue that may crop up again in the future.

**Comment by acammack-r7** [comment]
Rails has lots of ways to validate and introspect things before they are saved, but that's really the issue here. It looks like we aren't de-serializing the params properly, and maybe haven't ever: https://github.com/rapid7/metasploit-framework/blame/master/lib/msf/core/db_manager/import/metasploit_framework/xml.rb#L208-L210. Given the newline and spaces in the string we try to save, my guess is the YAML isn't being generated and/or parsed properly.

**Comment by gwillcox-r7** [comment]
# Release Notes
The XML data parser for the `db_import` command has been updated to fix a potential `nil` stack trace by implementing more appropriate handling of XML nodes. Additionally, a fix has been applied to ensure that `<body>` tags nested within a  `<web_page>` HTML tag are now appropriately Base64 decoded before being deserialized.

**Pull request merged**